### PR TITLE
Add `SharedUnionFieldsDeep` type

### LINKED ISSUE #750: Proposal: get the public part of union.
Hi everyone~
I'm having some problems with `OmitDeep` implementation, I want a type that can get the public part object type from union.

example:
```ts
type Union = { a: string, b: {b1: number, foo: string}, c: boolean } |
             { a: string, b: {b1: number, bar: number}, d: boolean }

type Result = PublicOfUnion<Union> // or other name
//=> type Result = { a: string, b: { b1: number } }
```

If the type is feasible, I'll try to implementation it but haven't idea now.

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/sindresorhus) so you can upvote and help fund this issue.
- The funding will be given to active contributors.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/sindresorhus/type-fest/issues/750">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/sindresorhus/type-fest/issues/750/pledge.svg?darkmode=1">
  Fund with Polar
</picture>
</a>

### PR DESCRIPTION
**Pull request by Emiyaaaaa**
- Closed #750

### CODE REVIEW

**Comment by sindresorhus** [comment]
What happened to the non-deep version that was discussed in https://github.com/sindresorhus/type-fest/issues/750 ?

File: source/internal.d.ts [hunk 1]
```diff
@@ -328,3 +356,73 @@ IsPrimitive<Object>
...
+
+Note: Just supports numbers from 0 to 999.
+
+@example
+```
+type A = MinNumbers<3 | 1 | 2>
+//=> 1
+```
+*/
+export type MinNumbers<N extends Number, T extends UnknownArray = []> =
```
Comment (sindresorhus) [inline source/internal.d.ts:407]: The name could be more specific. Maybe `UnionMin`?

Comment (sindresorhus) [inline source/internal.d.ts:407]:
```suggestion
export type MinNumbers<N extends number, T extends UnknownArray = []> =
```

Comment (sindresorhus) [reply source/internal.d.ts:407]: Should `T` really be exposed? May make more sense to wrap it in an internal type so this parameter is not exposed.

File: source/internal.d.ts [hunk 2]
```diff
@@ -328,3 +356,73 @@ IsPrimitive<Object>
...
+		: never; // Should never happen
+
+/**
+Returns the minimum number of the given union numbers.
+
+Note: Just supports numbers from 0 to 999.
+
+@example
+```
+type A = MinNumbers<3 | 1 | 2>
```
Comment (sindresorhus) [inline source/internal.d.ts:403]:
```suggestion
type A = MinNumbers<3 | 1 | 2>;
```

File: source/internal.d.ts [hunk 3]
```diff
@@ -328,3 +356,73 @@ IsPrimitive<Object>
...
+*/
+export type NonFixedPartOfArray<T extends UnknownArray> =
+	T extends unknown
+		? T extends readonly [...FixedPartOfArray<T>, ...infer U]
+			? U
+			: []
+		: never; // Should never happen
+
+/**
+Returns the minimum number of the given union numbers.
```
Comment (sindresorhus) [inline source/internal.d.ts:397]:
```suggestion
Returns the minimum number in the given union of numbers.
```

File: source/shared-union-fields-deep.d.ts [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+	T extends readonly [...infer U] ?
+		IsReadonly extends true
+			? readonly [...U]
+			: [...U]
+		: T;
+
+/**
+Returns whether the given array `T` is readonly.
+*/
+type ArrayIsReadonly<T extends UnknownArray> = T extends unknown[] ? false : true;
```
Comment (sindresorhus) [inline source/shared-union-fields-deep.d.ts:30]:
```suggestion
type IsArrayReadonly<T extends UnknownArray> = T extends unknown[] ? false : true;
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

**Comment by Emiyaaaaa** [comment]
> What happened to the non-deep version that was discussed in #750 ?

I want a deep version in #750 in fact, I can go non-deep version next step.

File: readme.md [hunk 5]
```diff
@@ -179,6 +179,7 @@ Click the type names for complete docs.
 - [`ArrayValues`](source/array-values.d.ts) - Provides all values for a constant array or tuple.
 - [`SetFieldType`](source/set-field-type.d.ts) - Create a type that changes the type of the given keys.
 - [`Paths`](source/paths.d.ts) - Generate a union of all possible paths to properties in the given object.
+- [`SharedUnionFieldsDeep`](source/shared-union-fields-deep.d.ts) - Create a deep version of anther object that has the shared fields of the given union object type.
```
Comment (sindresorhus) [inline readme.md:182]:
```suggestion
- [`SharedUnionFieldsDeep`](source/shared-union-fields-deep.d.ts) - Create a type with shared fields from a union of object types, deeply traversing nested structures.
```

File: source/internal.d.ts [hunk 6]
```diff
@@ -13,7 +14,34 @@ Infer the length of the given array `<T>`.
 
 @link https://itnext.io/implementing-arithmetic-within-typescripts-type-system-a1ef140a6f6f
 */
-type TupleLength<T extends readonly unknown[]> = T extends {readonly length: infer L} ? L : never;
+type ArrayLength<T extends readonly unknown[]> = T extends {readonly length: infer L} ? L : never;
+
+/**
+Infer the length of the given tuple `<T>`.
+
+Returns `never` if the given type is an non-fixed-length array like `Array<string>`.
```
Comment (sindresorhus) [inline source/internal.d.ts:22]: Isn't `non-fixed-length` just another name for a tuple?

File: source/internal.d.ts [hunk 7]
```diff
@@ -328,3 +356,73 @@ IsPrimitive<Object>
...
+Returns the fixed-indexed part of the given array.
+
+@example
+```
+type A = [string, number, boolean, ...string[]];
+type B = FixedPartOfArray<A>;
+//=> [string, number, boolean]
+```
+*/
+export type FixedPartOfArray<T extends UnknownArray, Result extends UnknownArray = []> =
```
Comment (sindresorhus) [inline source/internal.d.ts:370]:
```suggestion
export type StaticPartOfArray<T extends UnknownArray, Result extends UnknownArray = []> =
```

?

File: source/internal.d.ts [hunk 8]
```diff
@@ -328,3 +356,73 @@ IsPrimitive<Object>
 ```
 */
 export type IsPrimitive<T> = [T] extends [Primitive] ? true : false;
+
+/**
+Returns the fixed-indexed part of the given array.
```
Comment (sindresorhus) [inline source/internal.d.ts:361]:
```suggestion
Returns the static, fixed-length portion of the given array, excluding variable-length parts.
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: source/internal.d.ts (hunk 6 above)
Comment (Emiyaaaaa) [reply source/internal.d.ts:22]: > Isn't `non-fixed-length` just another name for a tuple?

tuple length is fixed, array length is non-fixed. I want to emphasise that this array is non-fixed and not a tuple.

Comment (sindresorhus) [reply source/internal.d.ts:22]: Ok

File: source/internal.d.ts [hunk 9]
```diff
@@ -328,3 +356,83 @@ IsPrimitive<Object>
...
+Returns the non-fixed-indexed part of the given array.
+
+@example
+```
+type A = [string, number, boolean, ...string[]];
+type B = NonFixedPartOfArray<A>;
+//=> string[]
+```
+*/
+export type NonFixedPartOfArray<T extends UnknownArray> =
```
Comment (Emiyaaaaa) [inline source/internal.d.ts:389]: so according to the above, maybe `VariablePartOfArray` better? @sindresorhus

Comment (sindresorhus) [reply source/internal.d.ts:389]: 👍

**Pull request merged**